### ISSUE TITLE
[Feature] SparkClient.write_to_cache: direct write to Trainer data cache

### ISSUE THREAD

**Issue by Shekharrajak** [kubeflow/sdk#510]
### What you would like to be added?

Add a direct write path from SparkClient into the Trainer data cache via Arrow Flight, eliminating the intermediate Iceberg/Parquet write to object storage when Spark output is consumed by a subsequent training job.

Current world : 

1. Read and Transform: Spark client connects and transforms the DataFrame. (datafusion-comet ON/OFF)
2. Write to S3: Spark converts Arrow to Parquet and saves to Iceberg on S3.
3. Initialize Trainer: Trainer client points to the Iceberg metadata JSON on S3.
4. Cache Data: Cache workers read Parquet via Iceberg and load into memory as Arrow.
5. Stream to Workers: Training workers fetch the Arrow data via Arrow Flight do_get.

Problems:

* 3 I/O hops: Spark → S3 → cache worker → trainer
* Full feature dataset duplicated on object storage

After : 

1. Connect & Transform: Spark client connects and transforms the DataFrame. (datafusion-comet ON/OFF)
2. Push to Cache: Spark writes data directly to the cache head via Arrow Flight do_put.
3. Zero-Copy Storage: Cache workers load Arrow RecordBatches directly into an indexable memory table without serialization.
4. Initialize Trainer: Trainer client connects directly to the cache head endpoint. 
5. Stream to Workers: Training workers fetch the Arrow data via Arrow Flight do_get.

Benefits : 

•  1 network hop, zero intermediate storage
•  Zero format conversions (Arrow end to end)
•  No S3 write cost, no S3 bandwidth pressure
•  Cache warm-up in seconds, not minutes
•  Single Python API: write_to_cache + DirectCacheInitializer
•  No Iceberg catalog or metadata path management required

pseudocode : 

```
from kubeflow.spark import SparkClient
from kubeflow.trainer import TrainerClient, CustomTrainer
from kubeflow.trainer.types import DirectCacheInitializer, Initializer

spark_client = SparkClient()
spark = spark_client.connect(num_executors=50, ...)

df = spark.read.parquet("s3a://raw/").filter(...).withColumn(...)

# New API
spark_client.write_to_cache(
    df=df,
    cache_head="grpc://data-cache-head-svc:50051",
    schema_name="reviews",
    table_name="processed",
    num_cache_workers=8,
)

TrainerClient().train(
    initializer=Initializer(
        dataset=DirectCacheInitializer(
            head_endpoint="grpc://data-cache-head-svc:50051",
        )
    ),
    trainer=CustomTrainer(func=train_fn, num_nodes=32, ...),
)
```

This proposal is intentionally a starting point. Feedback, alternative designs, and naming improvements are very welcome. 

### Why is this needed?

Less data copy over network, zero intermediate storage and improve overall ML pipeline 

### Love this feature?

Give it a 👍 We prioritize the features with most 👍

**Comment by HarshPopat23** [comment]
Hi @Shekharrajak and @andreyvelich, I’m interested in working on this feature.

I reviewed the current SparkClient, Trainer `DataCacheInitializer`, the Kubeflow Data Cache proposal, and the Trainer cache implementation. The Spark → Trainer direct handoff is still a genuine gap, and I could not find an existing PR implementing `write_to_cache`.

Before starting an implementation, I noticed that this requires coordination between `kubeflow/sdk` and `kubeflow/trainer`:

* The cache head’s [`[do_put](https://github.com/kubeflow/trainer/blob/master/pkg/data_cache/src/head/head_service.rs)`](https://github.com/kubeflow/trainer/blob/master/pkg/data_cache/src/head/head_service.rs) is currently unimplemented.
* Worker [`[do_put](https://github.com/kubeflow/trainer/blob/master/pkg/data_cache/src/worker/worker_service.rs)`](https://github.com/kubeflow/trainer/blob/master/pkg/data_cache/src/worker/worker_service.rs) currently receives Iceberg file-assignment metadata (`file_paths` and `row_start_indexes`), rather than arbitrary Arrow RecordBatches.
* The existing [`[DataCacheInitializer](https://github.com/kubeflow/sdk/blob/main/kubeflow/trainer/types/types.py)`](https://github.com/kubeflow/sdk/blob/main/kubeflow/trainer/types/types.py) requires `metadata_loc` and provisions the cache during TrainJob initialization. A direct Spark write therefore needs a clear cache-provisioning and lifecycle contract.

My proposed phased approach is:

1. **Define the API and ownership model**

   * Start with a pre-provisioned cache in the same namespace.
   * Define a typed `CacheWriteTarget` and return a `CacheReference` containing the endpoint, schema/table identity and write metadata.
   * Agree whether Trainer should consume this through a new `DirectCacheInitializer`, an extension to `DataCacheInitializer`, or another cache-reference API.

2. **Add a direct-ingestion protocol to Trainer data cache**

   * Use the head for dataset registration, worker discovery and commit coordination.
   * Route Arrow RecordBatches directly to cache workers rather than sending all data through the head.
   * Keep direct batch ingestion distinct from the existing file-assignment `do_put` contract.
   * Add schema validation, backpressure, partial-write handling and commit/abort behavior.

3. **Implement distributed Spark writes**

   * Stream Arrow batches from Spark executor partitions; avoid `collect()`, `toPandas()` or driver-side materialization.
   * Include stable write, partition and attempt identifiers so Spark task retries are idempotent and do not duplicate records.

4. **Integrate with Trainer**

   * Pass the returned cache reference to Trainer without requiring an Iceberg metadata file.
   * Document cache ownership, readiness, cleanup and failure behavior.

5. **Testing**

   * Rust tests for direct ingestion and worker routing.
   * SDK unit tests for API/backend behavior.
   * A Kind E2E workflow: Spark transformation → direct cache write → distributed TrainJob read.
   * Failure tests for schema mismatch, worker unavailability, partial commits and retried Spark partitions.

Would this architecture match the intended direction? In particular:

1. Should the first MVP assume a pre-provisioned cache, or should `SparkClient.write_to_cache()` also provision the LeaderWorkerSet and Services?
2. Should direct ingestion be designed as an extension of the existing cache protocol, or as a separate ingestion endpoint?
3. Would you prefer a design/KEP update and small feasibility prototype before implementation PRs?

If this direction is accepted, I would be happy to prepare the detailed design and split the work into reviewable SDK and Trainer sub-issues.

cc @akshaychitneni
